Web log improvements

l2isbad opened the pull request:
https://github.com/firehol/netdata/issues/2235
Not ready.
@ktsaou your help needed (charts, descriptions, dimensions, etc 😄 )

Initial support for squid logs added.

```yaml
squid_log:
  name: 'squid'
  type: 'squid'   # <--- new option
  path: '/var/log/squid3/access.log'
```

ktsaou commented:
wow! where do I start?
let me install this to try it...

ktsaou commented:
ok,

I am writing randomly, as I see them:

1. we have to add netdata to the `squid` group, if it exists.

2. For responses we had 3 charts: `response_statuses`, `response_codes` and `detailed_response_codes`. For squid you have disabled the first. I see in the specs squid adds 2 more codes: `000` (unavailable) and a few `6xx`, which should be accounted to `bad`.

3. The squid `http_code` I think should be the `detailed_response_codes` we already had. Also it should be `stacked`, not `area` (`detailed_response_codes` is stacked).

4. `squid_code` needs lower priority (higher number than `http_code` to go below it)
5. `hier_code`, should have family `hierarchy`.

I am reading the specs of squid and I see this:

> The Squid result code is composed of several tags (separated by underscore characters) which describe the response sent to the client.

It also groups codes like this:

- delivery method: `TCP`, `UDP`, or `NONE`.
- handling options (optional): `CLIENT`, `IMS`, `ASYNC`, `SWAPFAIL`, `REFRESH`, `SHARED`, `REPLY`.
- type of object (optional) `NEGATIVE`, `STALE`, `OFFLINE`, `INVALID`, `FAIL`, `MODIFIED`, `UNMODIFIED`, `REDIRECT`.
- cache events (optional): `HIT`, `MEM`, `MISS`, `DENIED`, `NOFETCH`, `TUNNEL`.
- delivery errors (optional): `ABORTED`, `TIMEOUT`.

So, it would be nice we could have them in separate charts. You can keep the `squid_code` chart as an overview of the combinations. So, I suggest to have

- family `squid_cache`, with charts `handling_opts`, `object_types`, `cache_events`
- family `squid_transport`, with charts `transport_methods`, `transport_errors`.

Once you have these, I'll install it on a high performing server I have to see what descriptions should be given above the charts and what alarms should be defined.

Execellent job! Wow! Very nice info!

ktsaou commented:
The log specs: http://wiki.squid-cache.org/SquidFaq/SquidLogs

l2isbad commented:
@ktsaou here you go

ktsaou commented:
ok, I have installed it.

1. The config does not have the squid log file. I suggest to add 2: `/var/log/squid3/access.log` (debian/ubuntu) and `/var/log/squid/access.log` (gentoo)

2. The `DENIED` requests have zero timing always. This affects the minimum and the average timings reported. I suggest not to count them for timing.

3. Chart `squid_code` should have the family `squid_cache`.

4. `squid_method` should have the family `requests` and the units should be `requests/s`.

5. It would be nice if you could trim `HIER_` from all `HIER_*` dimensions in chart `hier_code` (You can give the stripped one as dimension name, to continue report them in full name to netdata).

6. families `squid_*` could loose the `squid_` part (there is point for this duplication, they are under `squid` after all).

ktsaou commented:
7. I have a few unmatched lines... Any idea how to find them?

ktsaou commented:
8. The total clients chart we have for web servers is missing from squid.

ktsaou commented:
9. The `bytes` chart should follow the rules of the `bandwidth` chart of web servers. (it is line chart under squid). Also, I see that even for web servers it is KB/s but it should be kilobits/s (it is bandwidth)

ktsaou commented:
10. The `timings` charts has units `seconds`. Are you sure? I think they are `milliseconds`.

ktsaou commented:
well, I think my point 6 is wrong. I need to add different text on top of each section/family. Now they are talking about web server log file. From the charts I understand you are adding new charts for squid, you are not re-using the web ones. If this is the case, then `squid_` should be prepended to all of them. If you are re-using the same charts, I would need to adapt the texts to something more generic. Your call...

ktsaou commented:
If, to support squid, you are duplicating the code inside web log plugin, it might be better to have them separate. Again your call. If the complexity increases too much, it might be better to split.

l2isbad commented:
> The config does not have the squid log file. I suggest to add 2: /var/log/squid3/access.log (debian/ubuntu) and /var/log/squid/access.log (gentoo)

added

> The DENIED requests have zero timing always. This affects the minimum and the average timings reported. I suggest not to count them for timing.

done

> Chart squid_code should have the family squid_cache.

done

> squid_method should have the family requests and the units should be requests/s.

done

> It would be nice if you could trim HIER_ from all HIER_* dimensions in chart hier_code (You can give the stripped one as dimension name, to continue report them in full name to netdata).

done

> The total clients chart we have for web servers is missing from squid.

added

> The bytes chart should follow the rules of the bandwidth chart of web servers. (it is line chart under squid). Also, I see that even for web servers it is KB/s but it should be kilobits/s (it is bandwidth)

done (for web servers too)


> The timings charts has units seconds. Are you sure? I think they are milliseconds.

Well, divisor was 1000. Anyway changed to ms.

> If this is the case, then squid_ should be prepended to all of them.

done

> If the complexity increases too much, it might be better to split.

There is no complexity at all. Same is true for duplicating (except few charts!).
Keep in mind that i am not a programmer. I try to learn something doing stuff here and there 😄 

> I have a few unmatched lines... Any idea how to find them?

i suggest to add before
https://github.com/l2isbad/netdata/blob/web_log_improvements/python.d/web_log.chart.py#L832

```python
self.error('UNMATCHED', row)
```

to see unmatched rows in the log file

@ktsaou

ktsaou commented:
timings should be reported in microseconds and have a divider to milliseconds.
Now this is reported which is wrong:

![image](https://cloud.githubusercontent.com/assets/2662304/26577349/cbd5683e-4534-11e7-8732-190ea8ce93ba.png)

ktsaou commented:
The bandwidth units should be changed from `kb/s` to `kilobits/s` (to follow the same with all the other charts)

ktsaou commented:
remove `squid_` from families.

ktsaou commented:
Unmatched line:

```
UNMATCHED 1496137509.025    209 192.168.13.50 TCP_MISS/503 2083 GET http://ipv6.msftncsi.com/ncsi.txt - HIER_DIRECT/2a02:587:8ff:25::d4cd:4dbb text/html
```

ktsaou commented:
bandwidth chart should be area, not line

ktsaou commented:
edited a response above - check them on github.

l2isbad commented:
> add squid_ to all squid charts

[Already was there](https://github.com/l2isbad/netdata/blob/web_log_improvements/python.d/web_log.chart.py#L22-L25). Or i don't understand what do you mean.

ktsaou commented:
> Already was there. Or i don't understand what do you mean.

My bad. I meant: add `squid_` to all squid chart **families**, so I can provide info texts specifically designed for them. Sorry...

l2isbad commented:
> to all squid chart families

It will look strange..

Actually i can add underscore to the end. This will do the trick (timings_)
![screenshot_20170530_211701](https://cloud.githubusercontent.com/assets/22274335/26582795/61c4df94-457d-11e7-85a1-01f3c7eae479.png)

ktsaou commented:
> It will look strange..

i will remove them at `dashboard_info.js`. So, put the whole of it to be descriptive and I'll set the right name there.

ktsaou commented:
You can do it yourself too, in this PR.

[At this point](https://github.com/ktsaou/netdata/blob/f800734d07234c85a40bb77157048807a96f4ef8/web/dashboard_info.js#L292-L292), add:

```json
    'web_log.squid_bandwidth': {
        title: 'bandwidth',
        info: 'Bandwidth of responses (<code>sent</code>) by squid. This chart may present unusual spikes, since the bandwidth is accounted at the time the log line is saved by the server, even if the time needed to serve it spans across a longer duration. We suggest to use QoS (e.g. <a href="http://firehol.org/#fireqos" target="_blank">FireQOS</a>) for accurate accounting of the server bandwidth.'
    },
```

ktsaou commented:
Shall I test it again or are there more commits comming?

l2isbad commented:
> Shall I test it again

yes!

ktsaou commented:
final touches:

1. at `squid_hier_code`, `squid_transport_methods`, `squid_code`, `squid_cache_events` and `squid_handling_opts` **units** should be `requests/s`

2. timings are still wrong. I get values in the range of millions. I think the divider should be `* 1000`.

3. I still frequently get min timing as zero... hm...  it reports 0-1 ms for the queries made by netdata to squid API and a few more... a solution could be to (instead of ignoring `DENIED`) to ignore timings for all requests that report it as zero.

ktsaou commented:
Shall I also test the old web log usage?

ktsaou commented:
Regarding timings, I see millions are also reported by squid:

```
1496150037.419 3293567 192.168.2.85 TCP_MISS/200 2157298 CONNECT outlook.office365.com:443 - HIER_DIRECT/40.101.77.98 -
1496150037.419 5429939 192.168.2.85 TCP_MISS/200 317548 CONNECT outlook.office365.com:443 - HIER_DIRECT/40.101.55.210 -
1496150038.002 899790 192.168.14.24 TCP_MISS/200 4920 CONNECT mtalk.google.com:5228 - HIER_DIRECT/64.233.167.188 -

```

ktsaou commented:
![image](https://cloud.githubusercontent.com/assets/2662304/26585144/f4980f14-4553-11e7-94c0-62ab3b67b979.png)

12 million milliseconds is a long time..., but anyway this is what squid reports...

ktsaou commented:
And I got it... these are `CONNECT` sessions, used by clients using online web mail sessions. So, yes the clients were connected for that much time.

ktsaou commented:
So, fix the `units` and exclude the zero timed responses and I think we are done.
Then, we can merge this and I'll add info text for each section later today.

ktsaou commented on web/dashboard_info.js:
```diff
@@ -289,6 +289,31 @@ netdataDashboard.menu = {
+        title: 'hierarchy'
+      },
+    'web_log.squid_transport': {
+        title: 'squid_transport'
```
loose `squid_`.

ktsaou replied:
add:

```js
info: 'Analysis for the protocol used to receive the requests. <code>TCP</code> is used by clients, <code>UDP</code> is used by the <code>ICP</code> port (usually 3130) and <code>HTCP</code> port (usually 4128). <code>NONE</code> is used in all other cases (unusual response, or no response at all - it is also used in certain cases that the client requested a <code>CONNECT</code>)'
```

ktsaou commented on web/dashboard_info.js:
```diff
@@ -289,6 +289,31 @@ netdataDashboard.menu = {
+        title: 'squid_transport'
+      },
+    'web_log.squid_cache': {
+        title: 'squid_cache'
```
loose `squid_`

ktsaou replied:
add:

```js
info: 'Performance metrics for the performance of the squid cache.'
```

ktsaou commented on web/dashboard_info.js:
```diff
@@ -289,6 +289,31 @@ netdataDashboard.menu = {
+        title: 'requests'
+      },
+    'web_log.squid_hierarchy': {
+        title: 'hierarchy'
```
```js
info: 'Performance metrics for the squid hierarchy used to serve the requests.'
```

ktsaou commented on web/dashboard_info.js:
```diff
@@ -289,6 +289,31 @@ netdataDashboard.menu = {
+        title: 'squid_cache'
+      },
+    'web_log.squid_timings': {
+        title: 'timings'
```
Add:

```js
info: 'Duration of squid requests. Unrealistic spikes may be reported, since squid logs the total time of the requests, when they complete. Especially for HTTPS, the clients get a tunnel from the proxy and exchange requests directly with the upstream servers, so squid cannot evaluate the individual requests and reports the total time the tunnel was open.'
```

ktsaou commented:
I added a few info lines to submenus. I'll add more info to each chart later.
Let me know when you are ready to merge.

l2isbad commented:
?

squid_response_statuses - requests/s
squid_response_codes - responses/s (?)
squid_code - requests/s
squid_detailed_response_codes - responses/s (?)
squid_hier_code - requests/s
squid_method - requests/s
squid_transport_methods - requests/s
squid_transport_errors - requests/s
squid_handling_opts - requests/s
squid_object_types - requests/s
squid_cache_events - requests/s

ktsaou commented:
Did I mess it up?

I tried to understand which codes are used for stating the kind of request and which are used for responses.
I see I probably messed it up a bit.

squid_response_statuses - requests/s **should be `responses/s`**
squid_response_codes - responses/s (?) **it is ok**
squid_code - requests/s  **it is ok**
squid_detailed_response_codes - responses/s (?)  **it is ok**
squid_hier_code - requests/s  **it is ok**
squid_method - requests/s  **both could fit**
squid_transport_methods - requests/s  **it is ok**
squid_transport_errors - requests/s  **it is ok**
squid_handling_opts - requests/s **both could fit**
squid_object_types - requests/s **hm... `objects/s` ?**
squid_cache_events - requests/s **`events/s` ?**

l2isbad commented:
> Shall I also test the old web log usage?

Please do. (it works for me but still..)

> Did I mess it up?

looks fine for me but to be honest i didn't really go into specifics 😄

ktsaou commented:
web log seems good too.
I am merging it.

l2isbad commented:
@ktsaou 
https://github.com/firehol/netdata/blob/master/python.d/web_log.chart.py#L774-L783

`mime_type` extracted too. I'm not sure if this information is useful but requests per mime type sounds good.

?

ktsaou commented:
why not?
Add it and let's see if it provides something useful...
